# 07 - script22_output_comparison

This notebook __compares the pobtot__ (or any other given pop. col) __for every city__ in metro_gdf_2020 by using:

* f'censo_inegi_{year[2:]}_ageb'
* f'pobvoronoi_inegi_{year[2:]}_mzaageb_mza' #From Script 22
* f'pobvoronoi_inegi_{year[2:]}_mzaageb_node' #From Script 22
* f'pobvoronoi_inegi_{year[2:]}_mzaageb_hex' #From Script 22

__RESULTS:__
* 2010 comparison is a __success__ (Biggest diff is Laguna, -3.39 people)
* 2020 comparison is a __success__ (Diffs can be attributed to blocks 800, confirmed using Aguascalientes)

## Import libraries

In [1]:
main_folder_path = '../../../'

In [2]:
import os
import sys

import pandas as pd
import geopandas as gpd
import osmnx as ox
import numpy as np

from scipy.spatial import Voronoi, voronoi_plot_2d
import shapely

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

module_path = os.path.abspath(os.path.join(main_folder_path))
if module_path not in sys.path:
    sys.path.append(module_path)
    import aup
else:
    import aup

## Notebook config

In [3]:
# --------------- Database
# Cities
metro_schema = 'metropolis'
metro_table = 'metro_gdf_2020'
# Population data
pop_schema = 'sociodemografico'

## Load data

### Load all cities

In [4]:
metro_query = f"SELECT * FROM {metro_schema}.{metro_table}"
metro_gdf = aup.gdf_from_query(metro_query, geometry_col='geometry')

# Show
print(metro_gdf.shape)
metro_gdf.head(2)

(367, 6)


CVEGEO CVE_ENT CVE_MUN          NOMGEO   
0  01001      01     001  Aguascalientes  \
1  01005      01     005     Jesús María   

                                            geometry            city  
0  POLYGON ((-102.10641 22.06035, -102.10368 22.0...  Aguascalientes  
1  POLYGON ((-102.33568 22.05067, -102.33348 22.0...  Aguascalientes

## __2010 comparison__ [Success]

In [15]:
# Set comparison year (And tables)
year = 2010
# Population data - AGEBs
agebs_table = f'censo_inegi_{year[2:]}_ageb'
# Population data - Script 22 blocks
blocks_save_table = f'pobvoronoi_inegi_{year[2:]}_mzaageb_mza'
# Population data - Script 22 nodes
nodes_save_table = f'pobvoronoi_inegi_{year[2:]}_mzaageb_node'
# Population data - Script 22 hexs
hexs_save_table = f'pobvoronoi_inegi_{year[2:]}_mzaageb_hex'

# Prepare for comparison
comparison_df = pd.DataFrame()
i = 0
compare_col = 'pobtot'

# Iterate over each city loading different pop data and adding totals to comparison
city_lst = list(metro_gdf.city.unique()) #['Aguascalientes'] 
for city in city_lst:
    
     # --------------- SELECT CITY
    city_gdf = metro_gdf.loc[metro_gdf.city==city].copy()
    city_gdf = city_gdf.set_crs("EPSG:4326")

    # --------------- LOAD DATA

    # Load city's AGEBs
    print(f"{city} - Loading AGEBs.")
    # Create a tupple from a list with all unique cvegeo_mun ('CVE_ENT'+'CVE_MUN') of current city
    city_gdf['cvegeo_mun'] = city_gdf['CVE_ENT']+city_gdf['CVE_MUN']
    cvegeo_mun_lst = list(city_gdf.cvegeo_mun.unique())
    # To avoid error that happens when there's only one MUN in State: 
    # e.g.: <<< SELECT * FROM sociodemografico.censo_inegi_{year[2:]}_mza WHERE ("entidad" = '02') AND "mun" IN ('001',) >>>
    # Duplicate mun inside tupple if there's only one MUN.
    if len(cvegeo_mun_lst) >= 2:
        cvegeo_mun_tpl = str(tuple(cvegeo_mun_lst))
    else:
        cvegeo_mun_lst.append(cvegeo_mun_lst[0])
        cvegeo_mun_tpl = str(tuple(cvegeo_mun_lst))
    ageb_query = f"SELECT * FROM {pop_schema}.{agebs_table} WHERE \"cvegeo_mun\" IN {cvegeo_mun_tpl}"
    pop_ageb_gdf = aup.gdf_from_query(ageb_query, geometry_col='geometry')
    pop_ageb_gdf = pop_ageb_gdf.set_crs("EPSG:4326")

    # Load Script 22's blocks
    print(f"{city} - Loading Script 22's blocks.")
    blocks_query = f"SELECT * FROM {pop_schema}.{blocks_save_table} WHERE \"city\" LIKE \'{city}\'"
    s22_blocks_gdf = aup.gdf_from_query(blocks_query, geometry_col='geometry')

    # Load Script 22's nodes
    print(f"{city} - Loading Script 22's nodes.")
    nodes_query = f"SELECT * FROM {pop_schema}.{nodes_save_table} WHERE \"city\" LIKE \'{city}\'"
    s22_nodes_gdf = aup.gdf_from_query(nodes_query, geometry_col='geometry')

    # Load Script 22's hexs
    print(f"{city} - Loading Script 22's hexs.")
    hexs_query = f"SELECT * FROM {pop_schema}.{hexs_save_table} WHERE \"city\" LIKE \'{city}\'"
    s22_hexs_gdf = aup.gdf_from_query(hexs_query, geometry_col='geometry')
    s22_hexsres8_gdf = s22_hexs_gdf.loc[s22_hexs_gdf.res==8].copy()
    s22_hexsres9_gdf = s22_hexs_gdf.loc[s22_hexs_gdf.res==9].copy()
    s22_hexsres10_gdf = s22_hexs_gdf.loc[s22_hexs_gdf.res==10].copy()
    del s22_hexs_gdf

    # --------------- REGISTER DATA
    print(f"{city} - Registering Script 22's pop data.")
    comparison_df.loc[i,'city'] = city
    
    agebs_value = pop_ageb_gdf[compare_col].sum()
    comparison_df.loc[i,'agebs'] = agebs_value
    
    blocks_value = pop_ageb_gdf[compare_col].sum()
    comparison_df.loc[i,'blocks'] = blocks_value
    
    nodes_value = pop_ageb_gdf[compare_col].sum()
    comparison_df.loc[i,'nodes'] = nodes_value
    
    hexsres8_value = s22_hexsres8_gdf[compare_col].sum()
    comparison_df.loc[i,'hexsres8'] = hexsres8_value
    
    hexsres9_value = s22_hexsres9_gdf[compare_col].sum()
    comparison_df.loc[i,'hexsres9'] = hexsres9_value
    
    hexsres10_value = s22_hexsres10_gdf[compare_col].sum()
    comparison_df.loc[i,'hexsres10'] = hexsres10_value
    
    i+=1

Aguascalientes - Loading AGEBs.
Aguascalientes - Loading Script 22's blocks.
Aguascalientes - Loading Script 22's nodes.
Aguascalientes - Loading Script 22's hexs.
Aguascalientes - Registering Script 22's pop data.
Ensenada - Loading AGEBs.
Ensenada - Loading Script 22's blocks.
Ensenada - Loading Script 22's nodes.
Ensenada - Loading Script 22's hexs.
Ensenada - Registering Script 22's pop data.
Mexicali - Loading AGEBs.
Mexicali - Loading Script 22's blocks.
Mexicali - Loading Script 22's nodes.
Mexicali - Loading Script 22's hexs.
Mexicali - Registering Script 22's pop data.
Tijuana - Loading AGEBs.
Tijuana - Loading Script 22's blocks.
Tijuana - Loading Script 22's nodes.
Tijuana - Loading Script 22's hexs.
Tijuana - Registering Script 22's pop data.
La Paz - Loading AGEBs.
La Paz - Loading Script 22's blocks.
La Paz - Loading Script 22's nodes.
La Paz - Loading Script 22's hexs.
La Paz - Registering Script 22's pop data.
Los Cabos - Loading AGEBs.
Los Cabos - Loading Script 22's b

In [20]:
comparison_df['agebs-blocks'] = round(comparison_df['agebs']-comparison_df['blocks'],2)
comparison_df['agebs-nodes'] = round(comparison_df['agebs']-comparison_df['nodes'],2)
comparison_df['agebs-hexsres8'] = round(comparison_df['agebs']-comparison_df['hexsres8'],2)
comparison_df['agebs-hexsres9'] = round(comparison_df['agebs']-comparison_df['hexsres9'],2)
comparison_df['agebs-hexsres10'] = round(comparison_df['agebs']-comparison_df['hexsres10'],2)

# Show first 35
comparison_df.head(35)

city       agebs      blocks       nodes      hexsres8   
0   Aguascalientes    828378.0    828378.0    828378.0  8.283780e+05  \
1         Ensenada    398122.0    398122.0    398122.0  3.981220e+05   
2         Mexicali    838116.0    838116.0    838116.0  8.381158e+05   
3          Tijuana   1595559.0   1595559.0   1595559.0  1.595559e+06   
4           La Paz    225022.0    225022.0    225022.0  2.250220e+05   
5        Los Cabos    215545.0    215545.0    215545.0  2.155450e+05   
6         Campeche    233864.0    233864.0    233864.0  2.338640e+05   
7           Laguna   1066672.0   1066672.0   1066672.0  1.066675e+06   
8         Monclova    330608.0    330608.0    330608.0  3.306080e+05   
9   Piedras Negras    177291.0    177291.0    177291.0  1.772910e+05   
10        Saltillo    784671.0    784671.0    784671.0  7.846710e+05   
11          Colima    268875.0    268875.0    268875.0  2.688750e+05   
12       Tapachula    242821.0    242821.0    242821.0  2.428210e+05   
13          Tuxtla    652268.0    652268.0    652268.0  6.522680e+05   
14       Chihuahua    835009.0    835009.0    835009.0  8.350090e+05   
15        Delicias    164384.0    164384.0    164384.0  1.643840e+05   
16          Juarez   1324487.0   1324487.0   1324487.0  1.324487e+06   
17            CDMX   8810393.0   8810393.0   8810393.0  8.810384e+06   
18            ZMVM  10515550.0  10515550.0  10515550.0  1.051555e+07   
19         Durango    524266.0    524266.0    524266.0  5.242660e+05   
20          Celaya    514050.0    514050.0    514050.0  5.140500e+05   
21      Guanajuato    127018.0    127018.0    127018.0  1.270180e+05   
22            Leon   1452221.0   1452221.0   1452221.0  1.452221e+06   
23        Irapuato    433399.0    433399.0    433399.0  4.333991e+05   
24        Acapulco    699336.0    699336.0    699336.0  6.993360e+05   
25    Chilpancingo    248361.0    248361.0    248361.0  2.483610e+05   
26         Pachuca    422580.0    422580.0    422580.0  4.225800e+05   
27      Tulancingo    181998.0    181998.0    181998.0  1.819980e+05   
28     Guadalajara   4296874.0   4296874.0   4296874.0  4.296874e+06   
29        Vallarta    349200.0    349200.0    349200.0  3.492000e+05   
30          Piedad    139100.0    139100.0    139100.0  1.391000e+05   
31          Toluca   1599756.0   1599756.0   1599756.0  1.599756e+06   
32         Morelia    706804.0    706804.0    706804.0  7.068040e+05   
33          Zamora    225918.0    225918.0    225918.0  2.259176e+05   
34         Uruapan    294007.0    294007.0    294007.0  2.940070e+05   

        hexsres9     hexsres10  agebs-blocks  agebs-nodes  agebs-hexsres8   
0   8.283780e+05  8.283780e+05           0.0          0.0            0.04  \
1   3.981220e+05  3.981220e+05           0.0          0.0           -0.03   
2   8.381158e+05  8.381158e+05           0.0          0.0            0.24   
3   1.595559e+06  1.595559e+06           0.0          0.0           -0.00   
4   2.250220e+05  2.250220e+05           0.0          0.0           -0.02   
5   2.155450e+05  2.155450e+05           0.0          0.0            0.02   
6   2.338640e+05  2.338640e+05           0.0          0.0           -0.03   
7   1.066675e+06  1.066675e+06           0.0          0.0           -3.39   
8   3.306080e+05  3.306080e+05           0.0          0.0           -0.01   
9   1.772910e+05  1.772910e+05           0.0          0.0            0.05   
10  7.846710e+05  7.846710e+05           0.0          0.0            0.03   
11  2.688750e+05  2.688750e+05           0.0          0.0           -0.00   
12  2.428210e+05  2.428210e+05           0.0          0.0           -0.01   
13  6.522680e+05  6.522680e+05           0.0          0.0            0.01   
14  8.350090e+05  8.350090e+05           0.0          0.0           -0.03   
15  1.643840e+05  1.643840e+05           0.0          0.0            0.00   
16  1.324487e+06  1.324487e+06           0.0          0.0           -0.18   
17  8.810384e+06  8.810384e+06     

In [21]:
# Show last 36
comparison_df.tail(36)

city      agebs     blocks      nodes      hexsres8   
35         Cuautla   249448.0   249448.0   249448.0  2.494480e+05  \
36      Cuernavaca   884145.0   884145.0   884145.0  8.841450e+05   
37           Tepic   383232.0   383232.0   383232.0  3.832320e+05   
38       Monterrey  4131710.0  4131710.0  4131710.0  4.131710e+06   
39          Oaxaca   586495.0   586495.0   586495.0  5.864950e+05   
40          Puebla  2290801.0  2290801.0  2290801.0  2.290801e+06   
41      San Martin   232728.0   232728.0   232728.0  2.327280e+05   
42        Tehuacan   277745.0   277745.0   277745.0  2.777450e+05   
43       Queretaro   920993.0   920993.0   920993.0  9.209930e+05   
44          Cancun   674206.0   674206.0   674206.0  6.742060e+05   
45        Chetumal   182178.0   182178.0   182178.0  1.821780e+05   
46           Playa   155902.0   155902.0   155902.0  1.559020e+05   
47             SLP   994729.0   994729.0   994729.0  9.947291e+05   
48        Culiacan   746699.0   746699.0   746699.0  7.466990e+05   
49      Los Mochis   337066.0   337066.0   337066.0  3.370660e+05   
50        Mazatlan   406981.0   406981.0   406981.0  4.069810e+05   
51         Guaymas   171379.0   171379.0   171379.0  1.713790e+05   
52  Ciudad Obregon   378411.0   378411.0   378411.0  3.784110e+05   
53      Hermosillo   754918.0   754918.0   754918.0  7.549180e+05   
54         Nogales   215529.0   215529.0   215529.0  2.155290e+05   
55    Villahermosa   532801.0   532801.0   532801.0  5.328010e+05   
56        Victoria   305155.0   305155.0   305155.0  3.051550e+05   
57       Matamoros   456694.0   456694.0   456694.0  4.566940e+05   
58    Nuevo Laredo   380676.0   380676.0   380676.0  3.806760e+05   
59         Reynosa   695291.0   695291.0   695291.0  6.952906e+05   
60         Tampico   785986.0   785986.0   785986.0  7.859860e+05   
61        Tlaxcala   496833.0   496833.0   496833.0  4.968330e+05   
62   Coatzacoalcos   332276.0   332276.0   332276.0  3.322760e+05   
63         Cordoba   239697.0   239697.0   239697.0  2.396970e+05   
64      Minatitlan   245206.0   245206.0   245206.0  2.452060e+05   
65         Orizaba   349680.0   349680.0   349680.0  3.496800e+05   
66       Poza Rica   263597.0   263597.0   263597.0  2.635970e+05   
67        Veracruz   707669.0   707669.0   707669.0  7.076690e+05   
68          Xalapa   617603.0   617603.0   617603.0  6.176030e+05   
69          Merida   999376.0   999376.0   999376.0  9.993760e+05   
70       Zacatecas   303426.0   303426.0   303426.0  3.034260e+05   

        hexsres9     hexsres10  agebs-blocks  agebs-nodes  agebs-hexsres8   
35  2.494480e+05  2.494480e+05           0.0          0.0            0.00  \
36  8.841450e+05  8.841450e+05           0.0          0.0           -0.00   
37  3.832320e+05  3.832320e+05           0.0          0.0           -0.00   
38  4.131710e+06  4.131710e+06           0.0          0.0           -0.03   
39  5.864950e+05  5.864950e+05           0.0          0.0            0.00   
40  2.290801e+06  2.290801e+06           0.0          0.0           -0.02   
41  2.327280e+05  2.327280e+05           0.0          0.0           -0.01   
42  2.777450e+05  2.777450e+05           0.0          0.0           -0.01   
43  9.209930e+05  9.209930e+05           0.0          0.0            0.02   
44  6.742060e+05  6.742060e+05           0.0          0.0            0.03   
45  1.821780e+05  1.821780e+05           0.0          0.0            0.01   
46  1.559020e+05  1.559020e+05           0.0          0.0            0.01   
47  9.947291e+05  9.947291e+05           0.0          0.0           -0.05   
48  7.466990e+05  7.466990e+05           0.0          0.0            0.03   
49  3.370660e+05  3.370660e+05           0.0          0.0           -0.01   
50  4.069810e+05  4.069810e+05           0.0          0.0            0.01   
51  1.713790e+05  1.713790e+05           0.0          0.0           -0.01   
52  3.784110e+05  3.784110e+05           0.0          0.0            0.01 

## __2020 comparison__ [Success]

In [5]:
# Set comparison year (And tables)
year = '2020'
# Population data - AGEBs
agebs_table = f'censo_inegi_{year[2:]}_ageb'
# Population data - Script 22 blocks
blocks_save_table = f'pobvoronoi_inegi_{year[2:]}_mzaageb_mza'
# Population data - Script 22 nodes
nodes_save_table = f'pobvoronoi_inegi_{year[2:]}_mzaageb_node'
# Population data - Script 22 hexs
hexs_save_table = f'pobvoronoi_inegi_{year[2:]}_mzaageb_hex'

# Prepare for comparison
comparison_df = pd.DataFrame()
i = 0
compare_col = 'pobtot'

# Iterate over each city loading different pop data and adding totals to comparison
city_lst = list(metro_gdf.city.unique()) #['Aguascalientes'] 
for city in city_lst:
    
     # --------------- SELECT CITY
    city_gdf = metro_gdf.loc[metro_gdf.city==city].copy()
    city_gdf = city_gdf.set_crs("EPSG:4326")

    # --------------- LOAD DATA

    # Load city's AGEBs
    print(f"{city} - Loading AGEBs.")
    # Create a tupple from a list with all unique cvegeo_mun ('CVE_ENT'+'CVE_MUN') of current city
    city_gdf['cvegeo_mun'] = city_gdf['CVE_ENT']+city_gdf['CVE_MUN']
    cvegeo_mun_lst = list(city_gdf.cvegeo_mun.unique())
    # To avoid error that happens when there's only one MUN in State: 
    # e.g.: <<< SELECT * FROM sociodemografico.censo_inegi_{year[2:]}_mza WHERE ("entidad" = '02') AND "mun" IN ('001',) >>>
    # Duplicate mun inside tupple if there's only one MUN.
    if len(cvegeo_mun_lst) >= 2:
        cvegeo_mun_tpl = str(tuple(cvegeo_mun_lst))
    else:
        cvegeo_mun_lst.append(cvegeo_mun_lst[0])
        cvegeo_mun_tpl = str(tuple(cvegeo_mun_lst))
    ageb_query = f"SELECT * FROM {pop_schema}.{agebs_table} WHERE \"cvegeo_mun\" IN {cvegeo_mun_tpl}"
    pop_ageb_gdf = aup.gdf_from_query(ageb_query, geometry_col='geometry')
    pop_ageb_gdf = pop_ageb_gdf.set_crs("EPSG:4326")

    # Load Script 22's blocks
    print(f"{city} - Loading Script 22's blocks.")
    blocks_query = f"SELECT * FROM {pop_schema}.{blocks_save_table} WHERE \"city\" LIKE \'{city}\'"
    s22_blocks_gdf = aup.gdf_from_query(blocks_query, geometry_col='geometry')

    # Load Script 22's nodes
    print(f"{city} - Loading Script 22's nodes.")
    nodes_query = f"SELECT * FROM {pop_schema}.{nodes_save_table} WHERE \"city\" LIKE \'{city}\'"
    s22_nodes_gdf = aup.gdf_from_query(nodes_query, geometry_col='geometry')

    # Load Script 22's hexs
    print(f"{city} - Loading Script 22's hexs.")
    hexs_query = f"SELECT * FROM {pop_schema}.{hexs_save_table} WHERE \"city\" LIKE \'{city}\'"
    s22_hexs_gdf = aup.gdf_from_query(hexs_query, geometry_col='geometry')
    s22_hexsres8_gdf = s22_hexs_gdf.loc[s22_hexs_gdf.res==8].copy()
    s22_hexsres9_gdf = s22_hexs_gdf.loc[s22_hexs_gdf.res==9].copy()
    s22_hexsres10_gdf = s22_hexs_gdf.loc[s22_hexs_gdf.res==10].copy()
    del s22_hexs_gdf

    # --------------- REGISTER DATA
    print(f"{city} - Registering Script 22's pop data.")
    comparison_df.loc[i,'city'] = city
    
    agebs_value = pop_ageb_gdf[compare_col].sum()
    comparison_df.loc[i,'agebs'] = agebs_value
    
    blocks_value = s22_blocks_gdf[compare_col].sum()
    comparison_df.loc[i,'blocks'] = blocks_value
    
    nodes_value = s22_nodes_gdf[compare_col].sum()
    comparison_df.loc[i,'nodes'] = nodes_value
    
    hexsres8_value = s22_hexsres8_gdf[compare_col].sum()
    comparison_df.loc[i,'hexsres8'] = hexsres8_value
    
    hexsres9_value = s22_hexsres9_gdf[compare_col].sum()
    comparison_df.loc[i,'hexsres9'] = hexsres9_value
    
    hexsres10_value = s22_hexsres10_gdf[compare_col].sum()
    comparison_df.loc[i,'hexsres10'] = hexsres10_value
    
    i+=1

Aguascalientes - Loading AGEBs.
Aguascalientes - Loading Script 22's blocks.
Aguascalientes - Loading Script 22's nodes.
Aguascalientes - Loading Script 22's hexs.
Aguascalientes - Registering Script 22's pop data.
Ensenada - Loading AGEBs.
Ensenada - Loading Script 22's blocks.
Ensenada - Loading Script 22's nodes.
Ensenada - Loading Script 22's hexs.
Ensenada - Registering Script 22's pop data.
Mexicali - Loading AGEBs.
Mexicali - Loading Script 22's blocks.
Mexicali - Loading Script 22's nodes.
Mexicali - Loading Script 22's hexs.
Mexicali - Registering Script 22's pop data.
Tijuana - Loading AGEBs.
Tijuana - Loading Script 22's blocks.
Tijuana - Loading Script 22's nodes.
Tijuana - Loading Script 22's hexs.
Tijuana - Registering Script 22's pop data.
La Paz - Loading AGEBs.
La Paz - Loading Script 22's blocks.
La Paz - Loading Script 22's nodes.
La Paz - Loading Script 22's hexs.
La Paz - Registering Script 22's pop data.
Los Cabos - Loading AGEBs.
Los Cabos - Loading Script 22's b

In [6]:
comparison_df['agebs-blocks'] = round(comparison_df['agebs']-comparison_df['blocks'],2)
comparison_df['agebs-nodes'] = round(comparison_df['agebs']-comparison_df['nodes'],2)
comparison_df['agebs-hexsres8'] = round(comparison_df['agebs']-comparison_df['hexsres8'],2)
comparison_df['agebs-hexsres9'] = round(comparison_df['agebs']-comparison_df['hexsres9'],2)
comparison_df['agebs-hexsres10'] = round(comparison_df['agebs']-comparison_df['hexsres10'],2)

# Show first 35
comparison_df.head(35)

city       agebs      blocks         nodes      hexsres8   
0   Aguascalientes   1042295.0   1043905.0  1.043905e+06  1.043905e+06  \
1         Ensenada    397099.0    396709.0  3.967090e+05  3.967090e+05   
2         Mexicali    950970.0    959257.0  9.592567e+05  9.592567e+05   
3          Tijuana   2000259.0   2002469.0  2.002469e+06  2.002469e+06   
4           La Paz    270846.0    270681.0  2.706810e+05  2.706810e+05   
5        Los Cabos    338979.0    338675.0  3.386750e+05  3.386750e+05   
6         Campeche    267558.0    267387.0  2.673870e+05  2.673870e+05   
7           Laguna   1218759.0   1218686.0  1.218686e+06  1.218686e+06   
8         Monclova    364881.0    367434.0  3.674340e+05  3.674340e+05   
9   Piedras Negras    206172.0    206163.0  2.061630e+05  2.061630e+05   
10        Saltillo    996551.0    996503.0  9.965030e+05  9.965030e+05   
11          Colima    308946.0    308860.0  3.088600e+05  3.088600e+05   
12       Tapachula    262090.0    270748.0  2.707480e+05  2.707480e+05   
13          Tuxtla    742182.0    742075.0  7.420750e+05  7.420750e+05   
14       Chihuahua    968372.0    967934.0  9.679340e+05  9.679340e+05   
15        Delicias    179246.0    181747.0  1.817470e+05  1.817470e+05   
16          Juarez   1503616.0   1506851.0  1.506851e+06  1.506851e+06   
17            CDMX   9138524.0   9145155.0  9.145152e+06  9.145152e+06   
18            ZMVM  11712231.0  11814850.0  1.181485e+07  1.181485e+07   
19         Durango    623992.0    623988.0  6.239881e+05  6.239881e+05   
20          Celaya    577067.0    576257.0  5.762570e+05  5.762569e+05   
21      Guanajuato    148108.0    150237.0  1.502370e+05  1.502370e+05   
22            Leon   1774963.0   1775917.0  1.775917e+06  1.775917e+06   
23        Irapuato    507398.0    506892.0  5.068920e+05  5.068920e+05   
24        Acapulco    693789.0    690760.0  6.907600e+05  6.907600e+05   
25    Chilpancingo    297731.0    296004.0  2.960040e+05  2.960040e+05   
26         Pachuca    558026.0    557836.0  5.578360e+05  5.578360e+05   
27      Tulancingo    201513.0    201312.0  2.013120e+05  2.013120e+05   
28     Guadalajara   4999609.0   5011955.0  5.011955e+06  5.011955e+06   
29        Vallarta    444630.0    446298.0  4.462980e+05  4.462980e+05   
30          Piedad    149826.0    149709.0  1.497090e+05  1.497090e+05   
31          Toluca   1872266.0   1973234.0  1.973234e+06  1.973234e+06   
32         Morelia    857537.0    862378.0  8.623780e+05  8.623780e+05   
33          Zamora    249989.0    249853.0  2.498523e+05  2.498523e+05   
34         Uruapan    331860.0    334318.0  3.343180e+05  3.343180e+05   

        hexsres9     hexsres10  agebs-blocks  agebs-nodes  agebs-hexsres8   
0   1.043905e+06  1.043905e+06       -1610.0     -1609.98        -1609.97  \
1   3.967090e+05  3.967090e+05         390.0       389.98          389.98   
2   9.592567e+05  9.592567e+05       -8287.0     -8286.73        -8286.73   
3   2.002469e+06  2.002469e+06       -2210.0     -2210.03        -2210.03   
4   2.706810e+05  2.706810e+05         165.0       165.00          165.00   
5   3.386750e+05  3.386750e+05         304.0       304.00          304.00   
6   2.673870e+05  2.673870e+05         171.0       171.02          171.02   
7   1.218686e+06  1.218686e+06          73.0        73.07           73.07   
8   3.674340e+05  3.674340e+05       -2553.0     -2553.01        -2553.01   
9   2.061630e+05  2.061630e+05           9.0         9.00            9.00   
10  9.965030e+05  9.965030e+05          48.0        48.00           48.00   
11  3.088600e+05  3.088600e+05          86.0        86.00           86.00   
12  2.707480e+05  2.707480e+05       -8658.0     -8658.01        -8658.01   
13  7.420750e+05  7.420750e+05         107.0       107.00          107.00   
14  9.679340e+05  9.679340e+05         438.0       437.96          437.97   
15  1.817470e+05  1.817470e+05       -2501.0     -2500.99        -2501.00   
16  1.506851e+06  1.506851e+06       -32

In [7]:
# Show last 36
comparison_df.tail(36)

city      agebs     blocks         nodes      hexsres8   
35         Cuautla   269827.0   274340.0  2.743400e+05  2.743400e+05  \
36      Cuernavaca   983015.0   991976.0  9.919760e+05  9.919760e+05   
37           Tepic   438004.0   443405.0  4.434050e+05  4.434050e+05   
38       Monterrey  5245560.0  5245269.0  5.245269e+06  5.245269e+06   
39          Oaxaca   660807.0   662287.0  6.622870e+05  6.622870e+05   
40          Puebla  2654156.0  2647940.0  2.647940e+06  2.647940e+06   
41      San Martin   287426.0   285397.0  2.853970e+05  2.853970e+05   
42        Tehuacan   338779.0   337788.0  3.377880e+05  3.377880e+05   
43       Queretaro  1381392.0  1379350.0  1.379351e+06  1.379351e+06   
44          Cancun   928428.0   928339.0  9.283391e+05  9.283391e+05   
45        Chetumal   188996.0   188777.0  1.887770e+05  1.887770e+05   
46           Playa   327820.0   327652.0  3.276520e+05  3.276520e+05   
47             SLP  1194744.0  1194500.0  1.194500e+06  1.194500e+06   
48        Culiacan   886283.0   886053.0  8.860531e+05  8.860531e+05   
49      Los Mochis   378508.0   380961.0  3.809610e+05  3.809610e+05   
50        Mazatlan   473913.0   476405.0  4.764050e+05  4.764050e+05   
51         Guaymas   175083.0   175049.0  1.750490e+05  1.750490e+05   
52  Ciudad Obregon   408279.0   408279.0  4.082790e+05  4.082790e+05   
53      Hermosillo   904857.0   904842.0  9.048420e+05  9.048420e+05   
54         Nogales   261137.0   261136.0  2.611360e+05  2.611360e+05   
55    Villahermosa   576442.0   631715.0  6.317149e+05  6.317149e+05   
56        Victoria   332100.0   332080.0  3.320800e+05  3.320800e+05   
57       Matamoros   516075.0   516070.0  5.160699e+05  5.160699e+05   
58    Nuevo Laredo   423830.0   423830.0  4.238300e+05  4.238300e+05   
59         Reynosa   813143.0   813128.0  8.131281e+05  8.131281e+05   
60         Tampico   862030.0   861919.0  8.619190e+05  8.619190e+05   
61        Tlaxcala   586978.0   585297.0  5.852970e+05  5.852970e+05   
62   Coatzacoalcos   340360.0   340257.0  3.402570e+05  3.402570e+05   
63         Cordoba   241786.0   243712.0  2.437120e+05  2.437120e+05   
64      Minatitlan   244476.0   244042.0  2.440420e+05  2.440420e+05   
65         Orizaba   370978.0   378816.0  3.788160e+05  3.788160e+05   
66       Poza Rica   261251.0   266799.0  2.667990e+05  2.667990e+05   
67        Veracruz   800379.0   799476.0  7.994760e+05  7.994760e+05   
68          Xalapa   671742.0   683871.0  6.838710e+05  6.838710e+05   
69          Merida  1246526.0  1245902.0  1.245902e+06  1.245902e+06   
70       Zacatecas   369112.0   371409.0  3.714090e+05  3.714090e+05   

        hexsres9     hexsres10  agebs-blocks  agebs-nodes  agebs-hexsres8   
35  2.743400e+05  2.743400e+05       -4513.0     -4513.00        -4513.00  \
36  9.919760e+05  9.919760e+05       -8961.0     -8961.03        -8961.03   
37  4.434050e+05  4.434050e+05       -5401.0     -5401.00        -5401.01   
38  5.245269e+06  5.245269e+06         291.0       290.72          290.71   
39  6.622870e+05  6.622870e+05       -1480.0     -1480.02        -1480.01   
40  2.647940e+06  2.647940e+06        6216.0      6216.04         6216.05   
41  2.853970e+05  2.853970e+05        2029.0      2028.99         2028.99   
42  3.377880e+05  3.377880e+05         991.0       990.98          990.98   
43  1.379351e+06  1.379351e+06        2042.0      2041.01         2041.01   
44  9.283391e+05  9.283391e+05          89.0        88.89           88.88   
45  1.887770e+05  1.887770e+05         219.0       219.00          219.00   
46  3.276520e+05  3.276520e+05         168.0       168.03          168.03   
47  1.194500e+06  1.194500e+06         244.0       243.98          243.98   
48  8.860531e+05  8.860531e+05         230.0       229.93          229.93   
49  3.809610e+05  3.809610e+05       -2453.0     -2453.03        -2453.03   
50  4.764050e+05  4.764050e+05       -2492.0     -2491.97        -2491.97   
51  1.750490e+05  1.750490e+05          

### Find 2020 problem [Aguascalientes]

In [8]:
# Iterate over each city loading different pop data and adding totals to comparison
city_lst = ['Aguascalientes'] 
for city in city_lst:
    
     # --------------- SELECT CITY
    city_gdf = metro_gdf.loc[metro_gdf.city==city].copy()
    city_gdf = city_gdf.set_crs("EPSG:4326")

    # --------------- LOAD DATA

    # Load city's AGEBs
    print(f"{city} - Loading AGEBs.")
    # Create a tupple from a list with all unique cvegeo_mun ('CVE_ENT'+'CVE_MUN') of current city
    city_gdf['cvegeo_mun'] = city_gdf['CVE_ENT']+city_gdf['CVE_MUN']
    cvegeo_mun_lst = list(city_gdf.cvegeo_mun.unique())
    # To avoid error that happens when there's only one MUN in State: 
    # e.g.: <<< SELECT * FROM sociodemografico.censo_inegi_{year[2:]}_mza WHERE ("entidad" = '02') AND "mun" IN ('001',) >>>
    # Duplicate mun inside tupple if there's only one MUN.
    if len(cvegeo_mun_lst) >= 2:
        cvegeo_mun_tpl = str(tuple(cvegeo_mun_lst))
    else:
        cvegeo_mun_lst.append(cvegeo_mun_lst[0])
        cvegeo_mun_tpl = str(tuple(cvegeo_mun_lst))
    ageb_query = f"SELECT * FROM {pop_schema}.{agebs_table} WHERE \"cvegeo_mun\" IN {cvegeo_mun_tpl}"
    pop_ageb_gdf = aup.gdf_from_query(ageb_query, geometry_col='geometry')
    pop_ageb_gdf = pop_ageb_gdf.set_crs("EPSG:4326")

    # Load Script 22's blocks
    print(f"{city} - Loading Script 22's blocks.")
    blocks_query = f"SELECT * FROM {pop_schema}.{blocks_save_table} WHERE \"city\" LIKE \'{city}\'"
    s22_blocks_gdf = aup.gdf_from_query(blocks_query, geometry_col='geometry')

    # Load Script 22's nodes
    print(f"{city} - Loading Script 22's nodes.")
    nodes_query = f"SELECT * FROM {pop_schema}.{nodes_save_table} WHERE \"city\" LIKE \'{city}\'"
    s22_nodes_gdf = aup.gdf_from_query(nodes_query, geometry_col='geometry')

    # Load Script 22's hexs
    print(f"{city} - Loading Script 22's hexs.")
    hexs_query = f"SELECT * FROM {pop_schema}.{hexs_save_table} WHERE \"city\" LIKE \'{city}\'"
    s22_hexs_gdf = aup.gdf_from_query(hexs_query, geometry_col='geometry')
    s22_hexsres8_gdf = s22_hexs_gdf.loc[s22_hexs_gdf.res==8].copy()
    s22_hexsres9_gdf = s22_hexs_gdf.loc[s22_hexs_gdf.res==9].copy()
    s22_hexsres10_gdf = s22_hexs_gdf.loc[s22_hexs_gdf.res==10].copy()
    del s22_hexs_gdf

Aguascalientes - Loading AGEBs.
Aguascalientes - Loading Script 22's blocks.
Aguascalientes - Loading Script 22's nodes.
Aguascalientes - Loading Script 22's hexs.


In [10]:
pop_ageb_gdf.head(2)

nom_ent         nom_mun            nom_loc  pobtot  pobfem  pobmas   
0  Aguascalientes  Aguascalientes  Total AGEB urbana    2237  1137.0  1100.0  \
1  Aguascalientes  Aguascalientes  Total AGEB urbana    1411   712.0   699.0   

   p_0a2  p_0a2_f  p_0a2_m  p_3ymas  ...  cve_ent  cve_mun  cve_loc  cve_ageb   
0  216.0    103.0    113.0   2021.0  ...       01      001     0001      0017  \
1   45.0     24.0     21.0   1366.0  ...       01      001     0001      006A   

   cve_mza  cvegeo_mun  cvegeo_loc    cvegeo_ageb        cvegeo_mza   
0      000       01001   010010001  0100100010017  0100100010017000  \
1      000       01001   010010001  010010001006A  010010001006A000   

                                            geometry  
0  MULTIPOLYGON (((-102.25467 21.94536, -102.2549...  
1  MULTIPOLYGON (((-102.28214 21.94904, -102.2793...  

[2 rows x 235 columns]

In [16]:
# Sum blocks'pop by ageb
grouped_blocks = s22_blocks_gdf.groupby('cvegeo_ageb').agg({'pobtot':np.sum})
grouped_blocks.reset_index(inplace=True)
grouped_blocks.rename(columns={'pobtot':'blockpop'},inplace=True)

# Merge with ageb data
merged = pd.merge(pop_ageb_gdf[['pobtot','cvegeo_ageb']],grouped_blocks[['blockpop','cvegeo_ageb']],on='cvegeo_ageb')

# Find population differences
merged['pop_diff'] = merged['pobtot']-merged['blockpop']
merged.loc[merged.pop_diff!=0]

pobtot    cvegeo_ageb  blockpop  pop_diff
101    2001  0100100010888      1999         2
242    5275  0100100013721      5266         9
276    1237  0100100014113      1220        17
307     807  0100102374594       763        44
311    2351  0100102933929      2276        75
317    1501  0100103674607      1464        37
318    1115  0100103674611      1025        90
327    1998  010010237458A      1976        22
342    5218  0100500010497      4561       657
353    1638  0100500010656      1633         5
382    1354  0101100110193      1281        73
383    1248  0101100110206      1048       200

### NOTE: Analysis confirmed the three AGEBs with the greatest pop diff between pobtot (ageb data) and blockpop (script 22 data) comes from blocks 800.
* AGEB 100500010497 has a block 800 with 657 pobtot.
* AGEB 0101100110206 has a block 800 with 200 pobtot.
* AGEB 0100103674611 has a block 800 with 90 pobtot.